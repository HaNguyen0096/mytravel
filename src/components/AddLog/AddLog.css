@import '~antd/dist/antd.css';

.logForm{
  border-radius: 5px;
  background-color: #f2f2f2;
  padding: 20px;
}

textarea{
  display: block;
  width: 100%;
  padding: 10px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
}

button[type=submit]{
  width: 100%;
  background-color: #4CAF50;
  color: white;
  padding: 10px 20px;
  margin: 8px 0;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

button[type=submit]:hover {
  background-color: #45a049;
}

.visitDate{
  display: block;
  width: 150px;
  margin-top: 8px;
}

.privateBox{
  margin-top: 8px;
}

.setBox{
  margin-top: 8px;
  margin-bottom: 8px;
}

label{
  display: block;
}