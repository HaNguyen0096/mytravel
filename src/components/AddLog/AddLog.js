import React, { Component } from 'react'
import LogsContext from '../../contexts/LogsContext'
import ApiService from '../../services/api'
import ValidationError from '../ValidationError/ValidationError'
import { Button, Textarea } from '../Utils/Utils'
import {DatePicker} from 'antd'
import { Link } from 'react-router-dom'
import './AddLog.css'


export default class AddLog extends Component {

  constructor(props){
    super(props)
      this.state = {
        visit_day: '',
        latitude: '',
        longitude: '',
        title: '',
        description: '',
        image: '',
        rating: ''
      }
      this.onDateChange = this.onDateChange.bind(this)
  }

  updateLatitude(latitude){
    this.setState({latitude: latitude})
  }

  updateLongitude(longitude){
    this.setState({longitude: longitude})
  }

  updateTitle(title){
    this.setState({title: title})
  }

  updateDescription(description){
    this.setState({description: description})
  }

  updateImage(image){
    this.setState({image: image})
  }

  updateRating(rating){
    this.setState({rating: rating})
  }
  
  static contextType = LogsContext

  handleSubmit = ev => {
    ev.preventDefault()
    const { latitude, longitude, title, description, image, rating} = this.state
    const {setPrivate} = ev.target
    const lat = this.props.location ? this.props.location.latitude : latitude
    const lng = this.props.location ? this.props.location.longitude : longitude
    const visit_day = this.state.visit_day
    ApiService.postLog(lat, lng, title, description, image, rating, visit_day, setPrivate.value)
      .then(this.context.addLog)
      .catch(this.context.setError)
    this.props.location && this.props.onClose()
  }

  onDateChange = (val) => {
    const visit_day = val.toJSON().toString()
    this.setState({visit_day: visit_day})
  }

  validateLatitude(){
    const lat = this.state.latitude
    if (isNaN(lat) === true){
      return 'Please Enter A Number Between -90 and 90!'
    }
    else if (lat > 90 || lat < -90){
      return 'Invalid Latitude'
    }
  }

  validateLongitude(){
    const lng = this.state.longitude
    if (isNaN(lng) === true){
      return 'Please Enter A Number Between -90 and 90!'
    }
    else if (lng > 90 || lng < -90){
      return 'Invalid Longitude'
    }
  }

  validateRating(){
    const rating = this.state.rating
    if (isNaN(rating) === true){
      return 'Please Enter A Number Between 1 And 10!'
    }
    else if (rating > 10 || rating < 0){
      return 'Invalid Rating'
    }
  }

  validateVisitedDay(){
    const visited_day = this.state.visit_day
    if (!visited_day&&this.state.rating){
      return 'Please Pick a Date'
    }
  }

  render() {
    const latitudeError = this.validateLatitude()
    const longitudeError = this.validateLongitude()
    const ratingError = this.validateRating()
    const visitDayError = this.validateVisitedDay()
    return (
      <form
        className='logForm'
        onSubmit= {this.handleSubmit}        
      >
        <div className='addLog'>
        {!this.props.location &&<div className='addCoordinate'>
            <label htmlFor='coordinate'>Enter the latitude and longitude of the location (number between -90 and 90):</label>
            <div className='addLatitude'>
              <Textarea
                required
                type='number'
                step='0.0000000001'
                aria-label='Latitude...'
                name='latitude'
                id='latitude'
                cols='30'
                rows='2'
                placeholder='insert latitude..'
                onChange={e => this.updateLatitude(e.target.value)}>
              </Textarea>
              <ValidationError message={latitudeError}/>
            </div>
            <div className='addLongitude'>
              <Textarea
                required
                type='number'
                step='0.0000000001'
                aria-label='longitude...'
                name='longitude'
                id='longitude'
                cols='30'
                rows='2'
                placeholder='insert longitude..'
                onChange={e => this.updateLongitude(e.target.value)}>
              </Textarea>
              <ValidationError message={longitudeError}/>
            </div>
          </div>}
          <div className='addTitle'>
            <label htmlFor='title'>Title:</label>
            <Textarea
              required
              aria-label='title...'
              name='title'
              id='title'
              cols='30'
              rows='2'
              placeholder='insert title..'
              onChange={e => this.updateTitle(e.target.value)}>
            </Textarea>
          </div>
          <div className='addDescription'>
            <label htmlFor='description'>Description:</label>
            <Textarea
              required
              aria-label='description...'
              name='description'
              id='description'
              cols='30'
              rows='2'
              placeholder='insert description..'
              onChange={e => this.updateDescription(e.target.value)}>
            </Textarea>
          </div>
          <div className='addImage'>
            <label htmlFor='image'>Image (Please insert a link to your image):</label>
            <Textarea             
              aria-label='image...'
              name='image'
              id='image'
              cols='30'
              rows='2'
              placeholder='insert image..'
              onChange={e => this.updateImage(e.target.value)}>
            </Textarea>
          </div>
          <div className='addRating'>
            <label htmlFor='rating'>Rating (Enter a number between 1 and 10):</label>
            <Textarea
              required
              type='number'
              aria-label='rating...'
              name='rating'
              id='rating'
              cols='30'
              rows='2'
              placeholder='insert rating..'
              onChange={e => this.updateRating(e.target.value)}>
            </Textarea>
            <ValidationError message={ratingError}/>
          </div>
          <div className='addDate'>
            <label htmlFor='visitDate'>Visited Date:</label>
            <DatePicker required={true} className='visitDate' onChange={this.onDateChange} name='visit_day'/>
            <ValidationError message={visitDayError}/>
          </div>
          <div className='privateBox'>
            <label htmlFor='setPrivate'>Set Private:</label>
            <select className='setBox' id='setPrivate' name='setPrivate' 
              required>            
              <option value='true'>Public</option>
              <option value='false'>Private</option>             
            </select> 
          </div>
        </div>
        <Button 
          type='submit' 
          className='postLogBtn'
          disabled={
            this.validateLatitude() ||
            this.validateLongitude() ||
            this.validateRating() ||
            this.validateVisitedDay()
          }
        >
          Add Log
        </Button>
      </form>
    )
  }
}
