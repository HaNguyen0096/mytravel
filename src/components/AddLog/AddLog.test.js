import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import AddLog from './AddLog'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('AddLog component', () => {

  it('renders AddLog without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                      <AddLog />
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI for AddLog as expected', () => {
    const tree = renderer
      .create(<BrowserRouter>
                <AddLog />
              </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

  it('renders AddLog by default', () => {
    const wrapper = shallow(<AddLog />)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})