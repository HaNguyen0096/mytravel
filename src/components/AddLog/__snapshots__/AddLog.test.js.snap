// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLog component renders AddLog by default 1`] = `
<form
  className="logForm"
  onSubmit={[Function]}
>
  <div
    className="addLog"
  >
    <div
      className="addCoordinate"
    >
      <label
        htmlFor="coordinate"
      >
        Enter the latitude and longitude of the location (number between -90 and 90):
      </label>
      <div
        className="addLatitude"
      >
        <Textarea
          aria-label="Latitude..."
          cols="30"
          id="latitude"
          name="latitude"
          onChange={[Function]}
          placeholder="insert latitude.."
          required={true}
          rows="2"
          step="0.0000000001"
          type="number"
        />
        <ValidationError />
      </div>
      <div
        className="addLongitude"
      >
        <Textarea
          aria-label="longitude..."
          cols="30"
          id="longitude"
          name="longitude"
          onChange={[Function]}
          placeholder="insert longitude.."
          required={true}
          rows="2"
          step="0.0000000001"
          type="number"
        />
        <ValidationError />
      </div>
    </div>
    <div
      className="addTitle"
    >
      <label
        htmlFor="title"
      >
        Title:
      </label>
      <Textarea
        aria-label="title..."
        cols="30"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="insert title.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addDescription"
    >
      <label
        htmlFor="description"
      >
        Description:
      </label>
      <Textarea
        aria-label="description..."
        cols="30"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="insert description.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addImage"
    >
      <label
        htmlFor="image"
      >
        Image (Please insert a link to your image):
      </label>
      <Textarea
        aria-label="image..."
        cols="30"
        id="image"
        name="image"
        onChange={[Function]}
        placeholder="insert image.."
        rows="2"
      />
    </div>
    <div
      className="addRating"
    >
      <label
        htmlFor="rating"
      >
        Rating (Enter a number between 1 and 10):
      </label>
      <Textarea
        aria-label="rating..."
        cols="30"
        id="rating"
        name="rating"
        onChange={[Function]}
        placeholder="insert rating.."
        required={true}
        rows="2"
        type="number"
      />
      <ValidationError />
    </div>
    <div
      className="addDate"
    >
      <label
        htmlFor="visitDate"
      >
        Visited Date:
      </label>
      <Picker
        className="visitDate"
        name="visit_day"
        onChange={[Function]}
      />
    </div>
    <div
      className="privateBox"
    >
      <label
        htmlFor="setPrivate"
      >
        Set Private:
      </label>
      <select
        className="setBox"
        id="setPrivate"
        name="setPrivate"
        required={true}
      >
        <option
          value="true"
        >
          Public
        </option>
        <option
          value="false"
        >
          Private
        </option>
      </select>
    </div>
  </div>
  <Button
    className="postLogBtn"
    type="submit"
  >
    Add Log
  </Button>
</form>
`;

exports[`AddLog component renders the UI for AddLog as expected 1`] = `
<form
  className="logForm"
  onSubmit={[Function]}
>
  <div
    className="addLog"
  >
    <div
      className="addCoordinate"
    >
      <label
        htmlFor="coordinate"
      >
        Enter the latitude and longitude of the location (number between -90 and 90):
      </label>
      <div
        className="addLatitude"
      >
        <textarea
          aria-label="Latitude..."
          className="Textarea "
          cols="30"
          id="latitude"
          name="latitude"
          onChange={[Function]}
          placeholder="insert latitude.."
          required={true}
          rows="2"
          step="0.0000000001"
          type="number"
        />
      </div>
      <div
        className="addLongitude"
      >
        <textarea
          aria-label="longitude..."
          className="Textarea "
          cols="30"
          id="longitude"
          name="longitude"
          onChange={[Function]}
          placeholder="insert longitude.."
          required={true}
          rows="2"
          step="0.0000000001"
          type="number"
        />
      </div>
    </div>
    <div
      className="addTitle"
    >
      <label
        htmlFor="title"
      >
        Title:
      </label>
      <textarea
        aria-label="title..."
        className="Textarea "
        cols="30"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="insert title.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addDescription"
    >
      <label
        htmlFor="description"
      >
        Description:
      </label>
      <textarea
        aria-label="description..."
        className="Textarea "
        cols="30"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="insert description.."
        required={true}
        rows="2"
      />
    </div>
    <div
      className="addImage"
    >
      <label
        htmlFor="image"
      >
        Image (Please insert a link to your image):
      </label>
      <textarea
        aria-label="image..."
        className="Textarea "
        cols="30"
        id="image"
        name="image"
        onChange={[Function]}
        placeholder="insert image.."
        rows="2"
      />
    </div>
    <div
      className="addRating"
    >
      <label
        htmlFor="rating"
      >
        Rating (Enter a number between 1 and 10):
      </label>
      <textarea
        aria-label="rating..."
        className="Textarea "
        cols="30"
        id="rating"
        name="rating"
        onChange={[Function]}
        placeholder="insert rating.."
        required={true}
        rows="2"
        type="number"
      />
    </div>
    <div
      className="addDate"
    >
      <label
        htmlFor="visitDate"
      >
        Visited Date:
      </label>
      <div
        className="ant-picker visitDate"
        onMouseUp={[Function]}
      >
        <div
          className="ant-picker-input"
        >
          <input
            name="visit_day"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            placeholder="Select date"
            readOnly={true}
            size={12}
            title=""
            value=""
          />
          <span
            className="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              className="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      className="privateBox"
    >
      <label
        htmlFor="setPrivate"
      >
        Set Private:
      </label>
      <select
        className="setBox"
        id="setPrivate"
        name="setPrivate"
        required={true}
      >
        <option
          value="true"
        >
          Public
        </option>
        <option
          value="false"
        >
          Private
        </option>
      </select>
    </div>
  </div>
  <button
    className="Button postLogBtn"
    type="submit"
  >
    Add Log
  </button>
</form>
`;
